### PR TITLE
Github backend: Clear review tags on new patch

### PR DESCRIPTION
**Pull request by patrikfirek**

### CODE REVIEW

File: src/app/CodeRepo/lib/types/revision.js [hunk 1]
```diff
@@ -119,9 +119,18 @@ class Revision extends Type {
...
 
     async skipReview() {
-        this.tags = this.tags.filter((tag) => !tag.startsWith("review:"));
+        await this.clearTags("review:", false)
         this.tags.push(`review:skip`);
-        await this.save()
+        await this.save();
+    }
+
+    async clearTags(pattern, save=true) {
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/types/revision.js]: missing spaces around =
Should not this be a method on the Type-object so all types can benefit from it?

File: src/app/CodeRepo/lib/types/revision.js [hunk 2]
```diff
@@ -119,9 +119,18 @@ class Revision extends Type {
...
+        this.tags = this.tags.filter((tag) => !tag.startsWith(pattern));
+        if (save) {
+            await this.save;
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/types/revision.js]: Missing () after this.save

File: src/app/CodeRepo/lib/types/revision.js [hunk 3]
```diff
@@ -119,9 +119,18 @@ class Revision extends Type {
...
     async skipReview() {
-        this.tags = this.tags.filter((tag) => !tag.startsWith("review:"));
+        await this.clearTags("review:", false)
         this.tags.push(`review:skip`);
-        await this.save()
+        await this.save();
+    }
+
+    async clearTags(pattern, save=true) {
+        this.tags = this.tags.filter((tag) => !tag.startsWith(pattern));
```
Comment (mattiasrunge) [inline src/app/CodeRepo/lib/types/revision.js]: All changes to type objects should be "inplace" not replaced, thus splice must be used or we might end up with issues when different async methods work with the same object at in parallel

File: src/app/CodeRepo/lib/types/revision.js (hunk 1 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/types/revision.js]: done

File: src/app/CodeRepo/lib/types/revision.js (hunk 3 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/types/revision.js]: done

File: src/app/CodeRepo/lib/types/revision.js (hunk 2 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/types/revision.js]: done

File: src/app/CodeRepo/lib/types/revision.js [hunk 4]
```diff
@@ -119,9 +119,15 @@ class Revision extends Type {
...
-        this.tags = this.tags.filter((tag) => !tag.startsWith("review:"));
-        this.tags.push(`review:skip`);
-        await this.save()
+        await this.clearTags("review:", false);
+        this.tags.push("review:skip");
+        await this.save();
+    }
+
+    async clearReviews() {
+        this.reviews = [];
```
Comment (christensson) [inline src/app/CodeRepo/lib/types/revision.js]: Clear in-place. Splice can be used.

File: src/lib/typelib/lib/type.js [hunk 5]
```diff
@@ -273,7 +274,32 @@ class Type {
         }
 
         await this.save();
-        await notification.emit(`${this.constructor.typeName}.untagged`, this, tag);
+        await notification.emit(`${this.constructor.typeName}.untagged`, this, tags);
```
Comment (christensson) [inline src/lib/typelib/lib/type.js]: The thing emitted should be ensureArray(tags). Note that all subscribers to untagged must be updated to handle arrays.
Same thing needs to be done for tagged.

File: src/app/CodeRepo/lib/types/revision.js [hunk 6]
```diff
@@ -144,8 +150,8 @@ class Revision extends Type {
             });
         }
 
-        // Regenerate review tags
-        this.tags = this.tags.filter((tag) => !tag.startsWith("review:"));
+        // Clear and regenerate review tags
+        await this.clearTags("review:");
```
Comment (christensson) [inline src/app/CodeRepo/lib/types/revision.js]: call with save = false

File: src/lib/typelib/lib/type.js [hunk 7]
```diff
@@ -255,12 +256,12 @@ class Type {
         await notification.emit(`${this.constructor.typeName}.tagged`, this, tag);
```
Comment (christensson) [inline src/lib/typelib/lib/type.js]: Needs to be emitted as array always.

File: src/lib/typelib/lib/type.js [hunk 8]
```diff
@@ -273,7 +274,32 @@ class Type {
         }
 
         await this.save();
-        await notification.emit(`${this.constructor.typeName}.untagged`, this, tag);
+        await notification.emit(`${this.constructor.typeName}.untagged`, this, tags);
+    }
+
+    async clearTags(pattern, save = true) {
```
Comment (christensson) [inline src/lib/typelib/lib/type.js]: Rename pattern to tagStartsWith or something similar....

File: src/lib/typelib/lib/type.js [hunk 9]
```diff
@@ -273,7 +274,32 @@ class Type {
...
+        if (!pattern) {
+            throw new Error("Can not clear tags with null pattern");
+        }
+
+        const matchingTags = this.tags.filter((tag) => tag.startsWith(pattern));
+        console.log(`matchin tags: ${matchingTags}`);
```
Comment (christensson) [inline src/lib/typelib/lib/type.js]: debug print

File: src/lib/typelib/lib/type.js [hunk 10]
```diff
@@ -273,7 +274,32 @@ class Type {
...
+        const indicesToRemove = matchingTags.map((match) => this.tags.indexOf(match));
+        console.log(`indices to remove: ${indicesToRemove}`);
```
Comment (christensson) [inline src/lib/typelib/lib/type.js]: debug print

File: src/app/CodeRepo/lib/types/revision.js (hunk 4 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/types/revision.js]: done

File: src/app/CodeRepo/lib/types/revision.js (hunk 6 above)
Comment (patrikfirek) [inline src/app/CodeRepo/lib/types/revision.js]: done

File: src/lib/typelib/lib/type.js (hunk 5 above)
Comment (patrikfirek) [inline src/lib/typelib/lib/type.js]: done

File: src/lib/typelib/lib/type.js (hunk 8 above)
Comment (patrikfirek) [inline src/lib/typelib/lib/type.js]: done

File: src/lib/typelib/lib/type.js (hunk 9 above)
Comment (patrikfirek) [inline src/lib/typelib/lib/type.js]: oops, done

File: src/lib/typelib/lib/type.js (hunk 10 above)
Comment (patrikfirek) [inline src/lib/typelib/lib/type.js]: oops

**Pull request merged**